include/deal.II/fe/fe_poly.templates.h

```diff
@@ -472,43 +457,75 @@ FE_Poly<PolynomialType, dim, spacedim>::fill_fe_subface_values(
                           make_array_view(output_data.shape_3rd_derivatives,
                                           k));
 
-      for (unsigned int k = 0; k < this->dofs_per_cell; ++k)
-        correct_third_derivatives(output_data,
-                                  mapping_data,
-                                  quadrature.size(),
-                                  k);
+      correct_third_derivatives(output_data,
+                                mapping_data,
+                                quadrature.size(),
+                                mapping);
     }
 }
 
 
 
+template <class PolynomialType, int dim, int spacedim>
+inline void
+FE_Poly<PolynomialType, dim, spacedim>::correct_hessians(
+  internal::FEValuesImplementation::FiniteElementRelatedData<dim, spacedim>
+    &output_data,
+  const internal::FEValuesImplementation::MappingRelatedData<dim, spacedim>
+    &                           mapping_data,
+  const unsigned int            n_q_points,
+  const Mapping<dim, spacedim> &mapping) const
+{
+  // If we have a Cartesian mapping we don't need to correct the derivatives
+  // because the second and higher derivatives of the mapping are zero.
+  const bool need_to_correct_hessians =
+    nullptr == dynamic_cast<const MappingCartesian<dim> *>(&mapping);
+
+  if (need_to_correct_hessians)
```

Comment:
I think that a more general condition for the need for correction would actually be that `mapping_data.jacobian_pushed_forward_grads` is the zero tensor. Of course, you can shortcut this test if the mapping is Cartesian, as you use, but it might be useful to also use the more general test in a second step. That would also capture the case where someone uses a `MappingQ` (as we almost always do!) on a mesh with Cartesian cells (as we *also* almost always do).